core/src/display_object/video.rs

```diff
@@ -0,0 +1,336 @@
+    /// treated as a keyframe regardless of it being flagged as one.
+    keyframes: BTreeSet<u32>,
+}
+
+#[derive(Clone, Debug, Collect)]
+#[collect(require_static)]
+pub enum VideoSource {
+    SWF {
+        /// The movie that defined this video stream.
+        movie: Arc<SwfMovie>,
+
+        /// The video stream definition.
+        streamdef: DefineVideoStream,
+
+        /// The H.263 bitstream indexed by video frame ID.
+        ///
+        /// Each frame consists of a start and end parameter which can be used
+        /// to reconstruct a reference to the embedded bitstream.
+        frames: BTreeMap<u32, (usize, usize)>,
+    },
+}
+
+impl<'gc> Video<'gc> {
+    /// Construct a Video object that is tied to a SWF file's video stream.
+    pub fn from_swf_tag(
+        movie: Arc<SwfMovie>,
+        streamdef: DefineVideoStream,
+        mc: MutationContext<'gc, '_>,
+    ) -> Self {
+        let source = GcCell::allocate(
+            mc,
+            VideoSource::SWF {
+                movie,
+                streamdef,
+                frames: BTreeMap::new(),
+            },
+        );
+
+        Video(GcCell::allocate(
+            mc,
+            VideoData {
+                base: Default::default(),
+                source,
+                stream: None,
+                decoded_frame: None,
+                object: None,
+                keyframes: BTreeSet::new(),
+            },
+        ))
+    }
+
+    /// Preload frame data from an SWF.
+    ///
+    /// This function yields an error if this video player is not playing an
+    /// embedded SWF video.
+    pub fn preload_swf_frame(&mut self, tag: VideoFrame, context: &mut UpdateContext<'_, 'gc, '_>) {
+        match (*self
+            .0
+            .write(context.gc_context)
+            .source
+            .write(context.gc_context))
+        .borrow_mut()
+        {
+            VideoSource::SWF {
+                movie,
+                streamdef: _streamdef,
+                frames,
+            } => {
+                let subslice = SwfSlice::from(movie.clone()).to_unbounded_subslice(tag.data);
+
+                if frames.contains_key(&tag.frame_num.into()) {
+                    log::warn!("Duplicate frame {}", tag.frame_num);
+                }
+
+                if let Some(subslice) = subslice {
+                    frames.insert(tag.frame_num.into(), (subslice.start, subslice.end));
+                } else {
+                    log::warn!("Invalid bitstream subslice on frame {}", tag.frame_num);
+                }
+            }
+        }
+    }
+
+    /// Seek to a particular frame in the video stream.
+    pub fn seek(self, context: &mut UpdateContext<'_, 'gc, '_>, mut frame_id: u32) {
+        let read = self.0.read();
+        let source = read.source;
+        let stream = if let Some(stream) = &read.stream {
+            stream
+        } else {
+            log::error!("Attempted to sync uninstantiated video stream!");
```

Comment:
This is just a note, not _directly_ related to the code: IIRC I tend to get this error once on playback start, even if it then proceeds to play the video flawlessly. I didn't look into exactly why...

Reply:
It's because of `instantiate_child`: it calls `apply_place_object` before `post_instantiation`. Since video player seeking is handled by a `PlaceObject` tag with a given `ratio`, every place object seeks the video *before* it has registered it's video stream.

It's technically valid to start a video stream off mid-stream (as long as it's a keyframe), although I have no idea if Flash Player supports this. I suppose the correct thing would be to have this code store the requested seek frame and then actually do the seek in `post_instantiate`.

Reply:
Ah, I see. Thanks for the explanation!

Reply:
I pushed code to allow pre-post-instantiation seeking. That should fix this issue.

Reply:
Great!